central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+
+	var (
+		indicatorLookups []*v1.Query
+		indicatorIds     []string
+		indicatorsMap    = map[string]*storage.ProcessIndicator{}
+	)
+
+	for _, val := range portProcesses {
+		if val.Process == nil {
+			log.Warnf("Got PLOP object without Process information, ignore: %+v", val)
+			continue
+		}
+
+		indicatorLookups = append(indicatorLookups,
+			search.NewQueryBuilder().
+				AddExactMatches(search.ContainerName, val.Process.ContainerName).
+				AddExactMatches(search.PodID, val.Process.PodId).
+				AddExactMatches(search.ProcessName, val.Process.ProcessName).
+				AddExactMatches(search.ProcessArguments, val.Process.ProcessArgs).
+				AddExactMatches(search.ProcessExecPath, val.Process.ProcessExecFilePath).
+				ProtoQuery())
+	}
+
+	indicatorsQuery := search.DisjunctionQuery(indicatorLookups...)
+	log.Debugf("Sending query: %s", indicatorsQuery.String())
+	indicators, err := ds.indicatorDataStore.SearchRawProcessIndicators(ctx, indicatorsQuery)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, val := range indicators {
+		key := fmt.Sprintf("%s %s %s %s %s",
+			val.GetContainerName(),
+			val.GetPodId(),
+			val.GetSignal().GetName(),
+			val.GetSignal().GetArgs(),
+			val.GetSignal().GetExecFilePath(),
+		)
+
+		// A bit of paranoia is always good
+		if old, ok := indicatorsMap[key]; ok {
+			log.Warnf("An indicator %s is already present, overwrite with %s",
+				old.GetId(), val.GetId())
+		}
+
+		indicatorsMap[key] = val
+		indicatorIds = append(indicatorIds, val.GetId())
+	}
+
+	return indicatorsMap, indicatorIds, nil
+}
+
+// OpenClosedPLOPs is a convenient type alias to use in PLOP normalization
+type OpenClosedPLOPs struct {
+	open   []*storage.ProcessListeningOnPortFromSensor
+	closed []*storage.ProcessListeningOnPortFromSensor
+}
+
+// normalizePLOPs
+//
+// In the batch of PLOP events there could be many open & close events for the
+// same combination of port, protocol, process. Find and squash them into a
+// single event.
+//
+// Open/close state will be calculated from the totall number of open/close
+// events in the batch, assuming every single open will eventually be followed
+// by close. In this way in the case of:
+// * Out-of-order events, we would be able to establish correct status
+// * Pairs split across two batches, the status will be correct after processing both batches
+// * Lost events, the status will be incorrect
+// Another alternative would be to set the status based on the final PLOP
+// event, which will produce the same results for case 2 and 3. But such
+// approach will produce incorrect status in the case 1 as well, so counting
+// seems to be more preferrable.
+func normalizePLOPs(
+	plops []*storage.ProcessListeningOnPortFromSensor,
+) []*storage.ProcessListeningOnPortFromSensor {
+
+	normalizedMap := map[string]OpenClosedPLOPs{}
+	normalizedResult := []*storage.ProcessListeningOnPortFromSensor{}
+
+	for _, val := range plops {
+		key := fmt.Sprintf("%d %d %s",
```

Comment:
same key comment on spaces

Reply:
See above response to similar comment.

Reply:
looks like this one is still built with spaces.

Reply:
Oops. Fixed now.